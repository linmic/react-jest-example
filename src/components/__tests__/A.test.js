// jest.mock('../B', () => 'B');

import React, { Component } from 'react';
import { render, shallow, mount } from 'enzyme';

describe('<A />', () => {
  let makeSubject;
  let makeShallowSubject;
  let makeRenderSubject;
  let params;
  let A;

  beforeEach(() => {
    jest.resetModules();
    params = {
      bar: jest.fn(),
    };

    A = require('../A').default;

    makeSubject = () => mount(<A {...params} />);
    makeShallowSubject = () => shallow(<A {...params} />)
    makeRenderSubject = () => render(<A {...params} />);
  });
  
  describe('snapshots', () => {
    // snapshot
    it('should match latest render snapshot', () => {
      const subject = makeRenderSubject();

      expect(subject.html()).toMatchSnapshot();
    });

    it('should match latest shallow snapshot', () => {
      const subject = makeShallowSubject();

      expect(subject.html()).toMatchSnapshot();
    });

    it('should match latest mount snapshot', () => {
      const subject = makeSubject();

      expect(subject.html()).toMatchSnapshot();
    });
  });

  it('should set state.count to 1 if foo() is called', () => {
    const subject = makeSubject();

    subject.instance().foo();
    expect(subject.state().count).toBe(1);
  });

  // event simulate with shallow
  it('should call bar props when .btn-bar clicked', () => {
    const subject = makeShallowSubject();
    
    subject.find('.btn-bar').simulate('click');
    expect(params.bar).toHaveBeenCalled();
  });

  // event simulate with mount
  it('should call foo when .btn-foo clicked', () => {
    const subject = makeSubject();
    
    subject.instance().foo = jest.fn();
    subject.update();
    subject.find('.btn-foo').simulate('click');
    expect(subject.instance().foo).toHaveBeenCalled();
  });

  // setState
  it('should display the <h1> when this.state.showTitle is true, and vice versa', () => {
    const subject = makeSubject();
    const titleEl = subject.find('h1');

    // default state
    expect(titleEl.prop('style')).toMatchObject({ display: 'none' });

    // the same works for setProps
    subject.setState({ showTitle: true });
    expect(titleEl.prop('style')).toBe(undefined);
  });
});
