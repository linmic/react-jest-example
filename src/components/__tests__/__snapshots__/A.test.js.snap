// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<A /> snapshots should match latest mount snapshot 1`] = `
<A
  bar={[Function]}
>
  <div>
    <h1
      style={
        Object {
          "display": "none",
        }
      }
    >
      title
    </h1>
    <button
      className="btn-foo"
      onClick={[Function]}
    >
      click me
    </button>
    <button
      className="btn-bar"
      onClick={[Function]}
    >
      click me
    </button>
    <B>
      <nav>
        <C
          id={0}
          key="0"
        >
          <a
            href="#"
          >
            I am link 
            0
            .
          </a>
        </C>
        <C
          id={1}
          key="1"
        >
          <a
            href="#"
          >
            I am link 
            1
            .
          </a>
        </C>
        <C
          id={2}
          key="2"
        >
          <a
            href="#"
          >
            I am link 
            2
            .
          </a>
        </C>
      </nav>
    </B>
  </div>
</A>
`;

exports[`<A /> snapshots should match latest render snapshot 1`] = `
<div>
  <h1
    style="display:none;"
  >
    title
  </h1>
  <button
    class="btn-foo"
  >
    click me
  </button>
  <button
    class="btn-bar"
  >
    click me
  </button>
  <nav>
    <a
      href="#"
    >
      I am link 0.
    </a>
    <a
      href="#"
    >
      I am link 1.
    </a>
    <a
      href="#"
    >
      I am link 2.
    </a>
  </nav>
</div>
`;

exports[`<A /> snapshots should match latest shallow snapshot 1`] = `
<div>
  <h1
    style={
      Object {
        "display": "none",
      }
    }
  >
    title
  </h1>
  <button
    className="btn-foo"
    onClick={[Function]}
  >
    click me
  </button>
  <button
    className="btn-bar"
    onClick={[Function]}
  >
    click me
  </button>
  <B />
</div>
`;
